source/nodejs/adaptivecards/src/card-elements.ts

```diff
@@ -7153,6 +7165,10 @@ export class AdaptiveCard extends ContainerWithActions {
         return "AdaptiveCard";
     }
 
+    hasCarousel(): boolean {
+        return (this.getItemCount() === 1 && this.getItemAt(0).getJsonTypeName() === "Carousel");
+    }
```

Comment:
I'm not a big fan of this method -- what happens if we add another singleton-type element?

Reply:
Agree with @paulcam206. Related to that, `get currentCarouselPageId()` in `SubmitActionBase` isn't used anywhere and should be removed. Per my below suggestion, this `hasCarousel` method shouldn't be necessary anyway.

Reply:
@dclaux,
my above comments should have been pushed a few hours ago, so it's out of order.

`currentCarousePageId()` is used in action handler in card-designer-surface.ts to get the pageId when submit action is triggered.

Reply:
Ok, `currentCarouselPageId()` is used exclusively for display purposes in the designer - it shouldn't be a public API available in the renderer SDK. Just move the logic directly into card-designer-surface (as below) and remove `currentCarouselPageId` from `SubmitActionBase`:

```typescript
                let carouselPageId: string | undefined = undefined;

                const root = action.getRootObject() as AdaptiveCard;

                if (root) {
                    for (let i = 0; i < root.getItemCount(); i++) {
                        let element = root.getItemAt(i);

                        if (element instanceof Carousel) {
                            carouselPageId = element.currentPageId;
                        }
                    }
                }

                if (carouselPageId) {
                    message += `\ncarousel page id: "${carouselPageId}"`;
                }

```

Reply:
@dclaux, thanks for the suggestion, i will update it.

Reply:
@dclaux addressed it in this [PR](https://github.com/microsoft/AdaptiveCards/pull/6728)
for some reasons, i can't add you to PR.

Reply:
I think it's because someone in the team revoke some of my permissions.